include/AdePT/core/AsyncAdePTTransport.cuh

```diff
@@ -820,12 +825,23 @@ void TransportLoop(int trackCapacity, int scoringCapacity, int numThreads, Track
       // *** Transport ***
       // ------------------
 
+      AllowFinishOffEventArray allowFinishOffEvent;
+      for (int i = 0; i < kMaxThreads; ++i) {
+        // if waiting for transport to finish, the last N particles may be finished on CPU
+        if (eventStates[i].load(std::memory_order_acquire) == EventState::WaitingForTransportToFinish) {
+          allowFinishOffEvent.flags[i] = lastNParticlesOnCPU;
+        } else {
+          allowFinishOffEvent.flags[i] = 0;
+        }
+      }
+
       // *** ELECTRONS ***
       {
         const auto [threads, blocks] = computeThreadsAndBlocks(particlesInFlight[ParticleType::Electron]);
         TransportElectrons<PerEventScoring><<<blocks, threads, 0, electrons.stream>>>(
             electrons.tracks, electrons.queues.currentlyActive, secondaries, electrons.queues.nextActive,
-            electrons.queues.leakedTracksCurrent, gpuState.fScoring_dev, returnAllSteps, returnLastStep);
+            electrons.queues.leakedTracksCurrent, gpuState.fScoring_dev, gpuState.stats_dev, allowFinishOffEvent,
```

Comment:
I think we should replace, after the review, this ugly multi-parameter interface with a whiteboard object